### PR TITLE
First two modules in a form of a huuuuuuge blob.

### PR DESCRIPTION
**Pull request by G1gg1L3s**
# How to run

All paths in arguments are relatively to the workspace root.

## pals1

Binaries:
1. Convert hex to base64:

    `cargo run --bin hex264 <hex>`

2. Fixed xor. Xors two hex strings. To run:

    `cargo run --bin xor <a> <b>`

3. Single-byte XOR cipher. Breaks xor cipher with single-byte key. To run:

    `cargo run --bin sb-xor <string>`

    For example:

    `cargo run --bin sb-xor 1b37373331363f78151b7f2b783431333d78397828372d363c78373e783a393b3736`

4. Detect single-character XOR. Finds line that is encrypted with xoring with single-byte. Takes file as an arg. For example:

    `cargo run --bin detect-sb-xor pals1/4.txt`

5. Implement repeating-key XOR. Takes key as an argument and plaintext into stdin. To run:

    `cargo run --bin rk-xor <key>`

6. Break repeating-key XOR. Takes file as an argument. Example:

   `cargo run --bin break-xor pals1/6.txt`

7. AES in ECB mode. Decrypts file encrypted in aes-ecb with a provided key. Usage:

    `cargo run --bin aes-dec <key> <file>`

    For example:

    `cargo run --bin aes-dec "YELLOW SUBMARINE" pals1/7.txt`

8. Detect AES in ECB mode. Takes file and finds the line that is encrypted in ECB mode. For example:

    `cargo run --bin detect-aes pals1/8.txt`

## hack general

1. _Encoding_ section of [general](https://cryptohack.org/challenges/general/).

    `cargo run --bin encoding`

2.  _Encoding Challenge_. It makes 100 decoding requests and then displays the flag.

    `cargo run --bin 13377`

3.  _XOR starter_, _XOR properties_ and _Favourite byte_.

    `cargo run --bin bunch-of-xor`

4.  _You either know, XOR you don't_ challenge.

    `cargo run --bin multi-xor`

5. `Lemur XOR`. It takes two pngs and xores them, saving new png to `out.png`.

    `cargo run --bin png hack_general/lemur.png hack_general/flag.png`.

6.  _Greatest Common Divisor_, _Extended GCD_ and `Modular Arithmetic 1` challenges.

    `cargo run --bin gcd`

7.  _Data Formats_ section:

```
cargo run --bin private hack_general/privacy_enhanced_mail.pem

cargo run --bin der hack_general/2048b-rsa-example-cert.der

cargo run --bin  transparency hack_general/transparency.pem
```

 ## pals2

1. Takes string and length of block and padds this string:

    `cargo run --bin pkcs7 <msg> <len>`

2. Decrypt file encoded in base64 with the key:

    ` cargo run --bin cbc-dec <key> <file of base64>`
    
    For a quick example you can run:
    
    `cargo run --bin cbc-dec "YELLOW SUBMARINE" pals2/10.txt`
    
3. _An ECB/CBC detection oracle_ challenge. Just put the number of test as many as you want.

    `cargo run --bin detec-oracle <number of tests>`

4. _Byte-at-a-time ECB decryption (Simple)_.

    `cargo run --bin byte-at-a-time`

5. 13th challenge _ECB cut-and-paste_. Just put any email you want:

    `cargo run --bin cut-and-paste <email>`

6. 14th challenge _Byte-at-a-time ECB decryption (Harder)_.  It may take some time so it's better to compile with `--release`

   `cargo run --release --bin byte-at-a-time-harder.rs`

7. 16th challenge _CBC bitflipping attacks_.

    `cargo run --bin cbc-bitflip`

## pals3

1.  17th challenge _The CBC padding oracle_.

    `cargo run --release --bin cbc-oracle`

2. _Break fixed-nonce CTR mode using substitutions_

    `cargo run --bin ch19-break-ctr-manually`

3. _Break fixed-nonce CTR statistically_ challenge. To run:

    `cargo run --bin ch20-break-ctr-statistically pals3/20.txt`

4. _Crack an MT19937 seed_. The program may take up to 15 minutes to finish because of internal sleep calls.

    `cargo run --bin ch22-crack-mt`

5. Cloning the internal state of the MT19937.

    `cargo run --bin ch23-untemp-mt`

6.  Cracking the MT based cipher with a 16 bit key. It has a lot of brute force loops and sleep calls, so this binary can work for a long time.

    `cargo run --bin ch24-mt-cipher`

## poly

Contains only tests. To run them
`cargo test`

### AES
Challenges in that are dedicated to AES.

- _Keyed Permutations_ - no code, the answer is `bijection`
- _Resisting Bruteforce_ - no code, the answer is `biclique` (or something, I don't remember).
- _Structure of AES_ - `cargo run --bin matrix`
- _Round Keys_ - `cargo run --bin add-round-key`
- _Confusion through Substitution_ - `cargo run --bin sub-bytes`
- _Diffusion through Permutation_ - `cargo run --bin mix`
- _Bringing It All Together_ - `cargo run --bin dec`

### Block ciphers modes

- _Block Cipher Mode Starter_ - this challenge I solved in browser. Just made ecnryption and decryption queries.

- _Passwords as Keys_ challenge. Bruteforce passwords to decrypt ciphertext:

  `cargo run --bin pass <ciphertext> <passwords>`

  To get the flag:

  `cargo run --bin 2fb76b63a7e6421271099bb31997065167852ca9f72928d4f38f967dec8c8b78 hack_aes/words`

### ECB

This section contains only one challenge "ECB Oracle", which is the same as in pals2/byte-at-a-time. To run:

`cargo run --bin ecb-oracle 2> /dev/null`

It has a lot of debug logs, so it's better to redirect them to `/dev/null`

### CBC
1. _ECB CBC WTF_ . To run:

    `cargo run --bin ecbcbcwtf`

2. _Flipping Cookie_. The bit-flipping attack. To run:

    `cargo run --bin flipping`

3. _Lazy CBC_.  It's and attack on key that is used as iv.

    `cargo run --bin lazy-cbc`
    
### OFB

This section has only one simple challenge, that uses the symetric property of stream ciphers: `E(K, E(K, P)) = P`. So to solve it I used web browser, and just copy pasted the flag. 

### CTR

1.  _Bean Counter_ challenge. To run:

     `cargo run --bin bean-counter`

    It will result in a `bean.png`

2.  _CTRIME_ challenge:

    `cargo run --bin ctrime`

3. _Stream of Consciousness_. Very similar to `pals3/ch20-break-ctr-statistically`

    `cargo run --bin consciousness`

### GCM

To solve the _GCM_ challenge:

`cargo run --bin gcm`

### Other Modes

1. _LOGON ZERO_. To run (this may take a while):

    `cargo run --bin zero`

2.  _Paper Plane_ challenge. It's padding oracle attack but on IGE.
   To run (this may take a huge amount of time):

    `cargo run --bin paper-plane`

### Other ciphers

1.  _Triple DES_:

    `cargo run --bin 3des`

2.  _Oh SNAP_. Attack on RC4 weak vectors. As like as previous binary, this program will take a lot of your time. To run:

    `cargo run --bin rc4`

### CODE REVIEW

**Comment by storojs72** [comment]
As a first common remarks, please make following improvements:
1) add detailed description of your work (it should contain information about proposed solution, testing details, some important notes for reviewers that would save time on digging in your code);
2) as a continuation of first point - add instructions how to run your code; 
3) add signing git commits (https://docs.github.com/en/github/authenticating-to-github/signing-commits).

**Comment by storojs72** [comment]
Took a short second look. @G1gg1L3s at frist, please, reorganize codebase to be a workspace with internal rust crates (`poly`, `hack_aes`, `pals1`, `pals2`, etc). And secondly, when PR is created, please write a detailed description, don't leave it empty (for current PR you can at least copy-paste the info - how to run your binaries from README). This is ultimately important especially for such big PRs with 99 files changed!

File: hack_aes/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+/target
+Cargo.lock
```
Comment (storojs72) [inline hack_aes/.gitignore:2]: Should we keep `Cargo.lock` file in gitignore for Rust command-line tools?

File: pals1/src/bin/xor.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+use pals1::{
+    hex::{decode, encode},
+    xor,
+};
+
+fn main() {
+    let mut args = std::env::args().skip(1);
+    let string1 = args.next().expect("Usage: xor <hex> <hex>");
+    let string2 = args.next().expect("Usage: xor <hex> <hex>");
+    assert!(
```
Comment (storojs72) [inline pals1/src/bin/xor.rs:10]: Should we may be use more idiomatic `assert_eq`?

**Review by storojs72 — CHANGES_REQUESTED** [review]
PALS1
Challenge 1 - OK
Challenge 2 - OK
Challenge 3 - OK (Do you like Sherlock Holmes stories?)

Will review other challenges later. The only general comment - add please more user friendly error handling for all your command line applications instead of just crashing for corner cases:
```
artem@artem:~/CossackLabs/internship/nazar/crypto$ cargo run --bin sb-xor ffffff3331363f78151b7f2b783400333d78397828372d363c78373e783a393bfff
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/sb-xor ffffff3331363f78151b7f2b783400333d78397828372d363c78373e783a393bfff`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "Odd length"', pals1/src/bin/sb-xor.rs:8:31
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

You can use printing error message and exiting with non-zero code for example.

Question: just out of curiosity what IDE do you use for development (it is not an IntellijIDEA as far as I don't see `.idea` in `gitignore`)?

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+/target
+Cargo.lock
+/.vscode
```
[the remark this answers is not in the document]
Comment (G1gg1L3s) [reply .gitignore:3]: Yep, something went wrong with my formatter

File: hack_aes/.gitignore (hunk 1 above)
Comment (G1gg1L3s) [reply hack_aes/.gitignore:2]: I actually didn't know that, I thought Cargo.lock is an artifact that you shouldn't include. And it turns out that it's true only for lib crates. Thanks.

File: pals1/src/bin/xor.rs (hunk 2 above)
Comment (G1gg1L3s) [reply pals1/src/bin/xor.rs:10]: Yes, that makes sense.

**Comment by G1gg1L3s** [comment]
I actually have never read Sherlock Holmes stories, it's just the first book that came to my mind :)
Personally, I'm not a big fan of IDE, I prefer vscodium with a bunch of plugins, that's why you can sometimes find `.vscode` folders.

Also, this process of pull requesting and reviewing is kinda new for me. I have seen how it is done before, but never actually done it by myself, so if I do something wrong, please let me know :)
And one more thing, should I go and just check every file for similar errors and fix them? Is it done this way?

**Comment by storojs72** [comment]
> And one more thing, should I go and just check every file for similar errors and fix them? Is it done this way?

If you mean iterating each `main` function and replace crashing on user-friendly exiting - I would suggest to do so. I understand that main purpose of this exercises is learning - but nevertheless, let's follow best software engineering practices.

File: pals1/src/bin/rk-xor.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+use pals1::{hex::encode, xor};
+use std::{
+    error::Error,
+    io::{self, Read},
+};
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let key = std::env::args().nth(1).ok_or("Usage: rk_xor <key>")?;
+    // read entire stdin
+    let buffer: Vec<_> = {
```
Comment (storojs72) [inline pals1/src/bin/rk-xor.rs:10]: Should we may be foresee unlocking by timer? Currently program terminates infinitely if nothing is presented in `stdin`

**Review by storojs72 — CHANGES_REQUESTED** [review]
PALS1
Challenge 1 - OK
Challenge 2 - OK
Challenge 3 - OK
Challenge 4 - OK
Challenge 5 - OK
Challenge 6 - OK
Challenge 7 - OK
Challenge 8 - OK

Remarks:
1) That's a logical continuation of previous suggestion for user-friendly exiting in case of some unexpected input. Command line tools (it is not only for Rust but for any programming language) should provide descriptive explanation of available flags. Common practice is including `--help` flag that prints some default message. For example consider `git --help`:
```
artem@artem:~/CossackLabs/internship/nazar/crypto$ git --help
usage: git [--version] [--help] [-C <path>] [-c <name>=<value>]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p | --paginate | -P | --no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           <command> [<args>]

These are common Git commands used in various situations:

start a working area (see also: git help tutorial)
   clone             Clone a repository into a new directory
   init              Create an empty Git repository or reinitialize an existing one

 . . .
```
That is much more better than:
```
artem@artem:~/CossackLabs/internship/nazar/crypto$ cargo run --bin detect-sb-xor -- --help
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/detect-sb-xor --help`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', pals1/src/bin/detect-sb-xor.rs:11:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I do not insist on manually writing parser for command line arguments, but strongly recommend to use some already available projects for Rust language (they are also available for any other language) that specifically designed to provide mentioned arguments handling.

2) If we use different file for checking it also crashes:
```
artem@artem:~/CossackLabs/internship/nazar/crypto$ cargo run --bin detect-sb-xor pals1/7.txt
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/detect-sb-xor pals1/7.txt`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "Invalid hex character"', pals1/src/bin/detect-sb-xor.rs:13:34
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
A good indicator for proper input validation is a "Monkey test" - when you put different random strings with special characters. Also usually some corners are tested, like empty input

File: pals1/src/bin/rk-xor.rs (hunk 4 above)
Comment (G1gg1L3s) [reply pals1/src/bin/rk-xor.rs:10]: Is it done this way? I mean, I have never seen programs that have such timer, except of course password entry.

Comment (storojs72) [reply pals1/src/bin/rk-xor.rs:10]: Take a look at [asynk-std](https://docs.rs/crate/async-std/1.5.0) Rust crate. May be it will give you some ideas!

**Review by storojs72 — COMMENTED** [review]
CryptoHack - general

Encodings - OK
Encodings Challenge - OK
XOR - OK
Mathematics - OK
Data Formats - OK

1) Minor. Thank you for using `clap` and making your applications more standardized. I would only point that for example for `general/encoding` program, you could provide 4 flags:
- ascii
- hex
- base 64
- bytes_and_bigints

and make your app configurable. For instance if we run it with `cargo run --bin encoding -- --hex` it should return just one string: `Hex = You_will_be_working_with_hex_strings_a_lot`. Because now your programs just provide `--help` and `--version` flags - which is out of the box in `clap` without actual configuration power.

2)  Interest thing about transparency. Would you like to dig more on how it works? We currently have a project which is closely related to transparency domain. There is an excellent [blog-post ](https://research.swtch.com/tlog) about internals of server that maintains public append-only log. May be you can read it and write a short report for the next call, if you have a time of course

File: pals2/src/bin/detec-oracle.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    let key = random_bytes(KEY_SIZE);
+    if is_cbc {
+        let iv = random_bytes(BLOCK_SIZE);
+        (aes_cbc_enc(buff, &key, &iv), Mode::Cbc)
+    } else {
+        (aes_ecb_enc(buff, &key), Mode::Ecb)
+    }
+}
+
+fn detelection_oracle() {
```
Comment (storojs72) [inline pals2/src/bin/detec-oracle.rs:32]: Did you mean `detection_oracle`?

File: pals2/src/bin/detec-oracle.rs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    let len = prefix + data.len() + suffix;
+    let mut buff = Vec::with_capacity(len);
+    buff.extend(RandomIterator.take(prefix));
+    buff.extend_from_slice(data);
+    buff.extend(RandomIterator.take(suffix));
+    let is_cbc: bool = thread_rng().gen();
+    let key = random_bytes(KEY_SIZE);
+    if is_cbc {
+        let iv = random_bytes(BLOCK_SIZE);
+        (aes_cbc_enc(buff, &key, &iv), Mode::Cbc)
```
Comment (storojs72) [inline pals2/src/bin/detec-oracle.rs:26]: A bit confusing syntax. Plus, your function name `aes_cbc_enc` already has information of what it intends to do - encrypt, using CBC mode. I would avoid using redundant `Mode::Cbc`. May be it is better to have `aes_enc` function that would allow some parametrized input - e.g. `encParameters` expressed as a separate trait, where you can specify the mode of encryption and configure data (for CBC we need and IV)

**Review by storojs72 — COMMENTED** [review]
PALS2

Challenge 9 - OK
Challenge 10 - OK
Challenge 11 - OK
Challenge 12 - OK
Challenge 13 - OK
Challenge 14 - OK
Challenge 15 - OK
Challenge 16 - OK

File: pals2/src/bin/detec-oracle.rs (hunk 6 above)
Comment (G1gg1L3s) [reply pals2/src/bin/detec-oracle.rs:26]: The only purpose of this enum is to return information about which encryption algorithm was used, so we can lately verify our guess.

File: pals2/src/bin/detec-oracle.rs (hunk 5 above)
Comment (G1gg1L3s) [reply pals2/src/bin/detec-oracle.rs:32]: Yep

**Comment by G1gg1L3s** [comment]
I'm definitely gonna take a look at that blog-post. That sounds super interesting

File: pals2/src/bin/detec-oracle.rs (hunk 6 above)
Comment (storojs72) [reply pals2/src/bin/detec-oracle.rs:26]: That is the case. Using `aes_cbc_enc` already gives all necessary information - that CBC mode of AES is going to be used. And additional `Mode::Cbc` parameter seems redundant, isn't it?

**Comment by storojs72** [comment]
PALS3

Challenge 17 - OK
Challenge 18 - OK
Challenge 19 - OK
Challenge 20 - OK
Challenge 21 - OK
Challenge 22 - OK
Challenge 23 - OK
Challenge 24 - OK

Now I'm sure you will definitely use cryptographic PRNG in any crypto-related task!

File: pals2/src/bin/detec-oracle.rs (hunk 6 above)
Comment (G1gg1L3s) [reply pals2/src/bin/detec-oracle.rs:26]: But our encryption_oracle function encrypts randomly in CBC or ECB and we don't know which mode is used, so we can not be sure about our guess. So, I return this enum just to check it later.

**Comment by storojs72** [comment]
Could you please rephrase / provide more explanations to the tasks according to the modules? I don't clearly understand what parts of `hack_aes` relate to tasks in `CryptoHack Block Ciphers`.

**Comment by G1gg1L3s** [comment]
Are you talking about the names of binaries?

**Comment by storojs72** [comment]
No, I mean names of tasks according to modules from your internship program

**Comment by G1gg1L3s** [comment]
As I get, you don't understand which of the tasks from hack_aes are related to the `CryptoHack Block Ciphers`? But they all. On the main page of cryptohack this section is called `Block Ciphers` but link to that section is `https://cryptohack.org/challenges/aes/`, that's why I called this crate `hack_aes`. Did i understand the question correctly?

**Comment by storojs72** [comment]
From your internship program:
```
Module 1
- pals1 (70% of tasks)
- hack intro (100% of tasks)
- hack general part 1 - first 9 tasks from 18 (30% of tasks)

Module 2
- pals2 (70% of tasks);
- pals3 (70% of tasks)
- hack general part 2 - last 9 tasks from 18 (60% of tasks)
- hack block ciphers (60% of tasks)

Module 3
- pals4 (60% of tasks);
- pals5 (60% of tasks);
- hack general (80% of tasks - but you really should have almost of them from previous modules);
- hack dh (60% of tasks).

Module 4
- pals 6 (60% of tasks);
- hack rsa (60% of tasks);
- hack ecc (60% of tasks);
```

In your PR - I saw:
- pals1 (100%);
- pals2 (100%);
- pals3 (100%);
- hack general (100%);
- hack block ciphers (that you called hack_aes, right?) - and I can't identify how much is done, since naming is different

**Comment by G1gg1L3s** [comment]
Yep, hack block ciphers is hack_aes. And as I understand, I didn't have to do all the tasks, just some of them? :)

File: hack_aes/src/bin/bean-counter.rs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        (version: "1.0")
+        (about: "Solution to cryptohack 'Bean Counter challenge' challenge. Saves result to the file.")
+        (@arg OUT: -o --output +takes_value "Path to save the resulted png (default is 'bean.png')")
+    )
+    .get_matches();
+    let out = matches.value_of("OUT").unwrap_or("bean.png");
+    let png = get_raw();
+    let png = try_with_len(&png, 0xd);
+    let mut file = std::fs::File::create(out).unwrap();
+    file.write_all(&png).unwrap();
```
Comment (storojs72) [inline hack_aes/src/bin/bean-counter.rs:62]: Sometimes you are unable to write into file, due to lack of permissions or read-only disk drives. Should we think about handling such errors more user-friendly than just `unwrap` and crashing the program?

**Review by storojs72 — CHANGES_REQUESTED** [review]
HACK GENERAL:

AES tasks - OK;
Block Cipher Modes tasks - OK;
ECB - OK;
CBC- OK;
OFB - OK;
CTR - OK;
GCM - OK;
Other modes - OK;
Other ciphers - OK;

Some remarks:

1) Did you use `nightly` version of rust toolchain for some programs? I encountered following error while running `matrix` task:
```
error[E0554]: `#![feature]` may not be used on the stable release channel
 --> hack_aes/src/lib.rs:1:1
  |
1 | #![feature(num_as_ne_bytes)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

```
But it compiles and normally runs if I use `nightly` instead of `stable` toolchain. If so, then please in future include such kind of information into README / description of PR since this is quite important. Usually it is better to use stable version of software instead of experimental - since last is often a subject of frequent breaking changes.

2) I would suggest to always use `unwrap` with caution. It's OK to us it in such place:
```
String::from_utf8(vec![0x20, 0x21, 0x22]).unwrap()
```
because you definitely know that no errors will occur. In other cases we should always consider better error handling strategies

File: hack_aes/src/bin/pass.rs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    for pass in buffer.lines() {
+        let pass = &pass?;
+        let key = md5::compute(pass).0;
+        let aes = Aes128::new(&key);
+        let mut plaintext = ciphertext.clone();
+        for block in plaintext.chunks_mut(BLOCK_LEN) {
+            aes.decrypt_block(block);
+        }
+        if let Ok(plaintext) = String::from_utf8(plaintext) {
```
[the remark this answers is not in the document]
Comment (G1gg1L3s) [reply hack_aes/src/bin/pass.rs:36]: Generally yes, but in this case we shouldn't. We are iterating over the passwords and trying to decrypt our text with a key that is derived from a password. Also we know that we should get some string in a form of `crypto{...}`, so we know that it's valid utf8, and we know that if we don't hit the valid utf8, it's definitely not the right key and not the right password.
But there is another issue in this program: if we didn't find any password, our program will print nothing. So, I'm going to change this.

File: hack_aes/src/bin/rc4.rs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        error: String,
+    }
+    let url = format!(
+        "http://aes.cryptohack.org/oh_snap/send_cmd/{}/{}/",
+        hex::encode(data),
+        hex::encode(nonce)
+    );
+
+    let res = loop {
+        if let Ok(res) = ureq::get(&url).call() {
```
[the remark this answers is not in the document]
Comment (G1gg1L3s) [reply hack_aes/src/bin/rc4.rs:52]: Yes, it should be fixed. I've included this loop because when program makes a lot of requests some of them start to fail, which crashes the program and all progress of cracking the flag. And this loop helps preventing it. Maybe I should add some parameter for maximum number of tries?

File: hack_aes/src/bin/bean-counter.rs (hunk 7 above)
Comment (G1gg1L3s) [reply hack_aes/src/bin/bean-counter.rs:62]: Yep, It's just something I missed when refactoring errors.

File: hack_aes/src/bin/sub-bytes.rs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        [251, 64, 182, 81],
+        [146, 168, 33, 80],
+        [199, 159, 195, 24],
+        [64, 80, 182, 255],
+    ];
+
+    hack_aes::sub_bytes(&mut state, &hack_aes::INV_SBOX);
+
+    let bytes = hack_aes::matrix_to_bytes(&state);
+    println!("{}", String::from_utf8(bytes).unwrap());
```
[the remark this answers is not in the document]
Comment (G1gg1L3s) [reply hack_aes/src/bin/sub-bytes.rs:19]: Ok, I will refactor it

File: hack_aes/src/bin/rc4.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (G1gg1L3s) [reply hack_aes/src/bin/rc4.rs]: As I understand, this is the same comment as a previous one? Should I remove this?

[the remark this answers is not in the document]
Comment (storojs72) [reply hack_aes/src/bin/rc4.rs]: Yes. this is an expected solution

[the remark this answers is not in the document]
Comment (storojs72) [reply hack_aes/src/bin/rc4.rs]: Yes. If you can avoid this somehow - it is great. Or you should introduce number of tries and get url in a loop not more than number of tries

**Review by storojs72 — APPROVED** [review]
Only very minor things are left - not related to correctness of tasks solving. So I think it should be approved
